### PR TITLE
ci/ipsec: Fix downgrade version retrieval

### PR DESCRIPTION
**Pull request by qmonnet**
Figuring out the right "previous patch release version number" to downgrade to in print-downgrade-version.sh turns out to be more complex than expected [0] [1] [2] [3].

This commit is an attempt to 1) fix issues with the current script and 2) overall make the script clearer, so we can avoid repeating these mistakes.

As for the fixes, there are two things that are not correct with the current version. First, we're trying to validate the existence of the tag to downgrade to, in case the script runs on top of a release preparation commit for which file VERSION has been updated to a value that does not yet contains a corresponding tag. This part of the script is actually OK, but not the way we call it in the IPsec workflow: we use `fetch-tags: true` but `fetch-depth: 0` (the default), and the two are not compatible, a shallow clone results in no tags being fetched.

To address this, we retrieve the tag differently: instead of relying on "fetch-tags" from the workflow, we call "git fetch" from the script itself, provided the preconditions are met (we only run it from a Git repository, if the "origin" remote is defined). If the tag exists, either locally or remotely, then we can use it. Otherwise, the script considers that it runs from a release preparation Pull Request, and decrements the patch release number.

The second issue is that we would return no value from the script if the patch release is zero. This is to avoid any attempt to find a previous patch release when working on a development branch. However, this logics is incorrect (it comes from a previous version of the script where we would always decrement the patch number). After the first release of a new minor version, it's fine to have a patch number at 0. What we should check instead is whether the version ends with `-dev`.

This commit brings additional changes for clarity: more comments, and a better separation between the "get latest patch release" and "get previous stable branch" cases, moving the relevant code to independent functions, plus better argument handling. We also edit the IPsec workflow to add some logs about the version retrieved. The logs should also display the script's error messages, if any, that are printed to stderr.

Sample output from the script:

    VERSION     Tag exists  Prevous minor   Previous patch release

    1.14.3      Y           v1.13           v1.14.3
    1.14.1      Y           v1.13           v1.14.1
    1.14.0      Y           v1.13           v1.14.0
    1.14.1-dev  N           v1.13           <error>
    1.15.0-dev  N           v1.14           <error>
    1.13.90     N           v1.12           v1.13.89  <- decremented
    2.0.0       N           <error>         <error>
    2.0.1       N           <error>         v2.0.0    <- decremented
    2.1.1       N           v2.0            v2.1.0    <- decremented

[0]: https://github.com/cilium/cilium/commit/56dfec2f1ac5
[1]: https://github.com/cilium/cilium/commit/4d7902f54a74
[2]: https://github.com/cilium/cilium/commit/5581963cbf94
[3]: https://github.com/cilium/cilium/commit/3803f539a740

Split diff view recommended to review the new version of print-downgrade-version.sh.

### CODE REVIEW

**Comment by qmonnet** [comment]
Tested in CI, in particular for the “downgrade to latest patch version” job:

- On `main`, the steps are skipped (no latest patch version on branch, so this is expected)
- On `v1.15`, current HEAD, we downgrade to `v1.15.0` (expected)
- On `v1.15`, with a test commit to change VERSION's content to `v1.15.1`, we decrement and downgrade to `v1.15.0` (expected)
- On `v1.15`, with a test commit to change VERSION's content to `v1.16.0`, we skip the remaining steps (no way to decrement from `v1.16.0` - this is the expected behaviour).

File: contrib/scripts/print-downgrade-version.sh
Comment (gentoo-root) [inline contrib/scripts/print-downgrade-version.sh]: Tip: `grep -x` allows you to drop `^` and `$`.

File: contrib/scripts/print-downgrade-version.sh
Comment (gentoo-root) [inline contrib/scripts/print-downgrade-version.sh]: > Just make sure the VERSION update comes in the last commit of your PRs.

Uhm, even though it will be the case most of the time, I don't think it's obvious enough for the rare case when the release manager pushes some fixup after the "prepare for release" commit.

How about we explicitly try to fetch the tag stored in the VERSION file? This should do the thing:

```
git fetch --depth=1 origin "+refs/tags/$(<VERSION):refs/tags/$(<VERSION)"
```

File: contrib/scripts/print-downgrade-version.sh
Comment (pchaigno) [reply contrib/scripts/print-downgrade-version.sh]: (But everyone knows what `^` and `$` mean. Who knows by heart what `-x` means? :slightly_smiling_face:)

**Review by bimmlerd — APPROVED** [review]

File: contrib/scripts/print-downgrade-version.sh
Comment (qmonnet) [reply contrib/scripts/print-downgrade-version.sh]: > Uhm, even though it will be the case most of the time, I don't think it's obvious enough for the rare case when the release manager pushes some fixup after the "prepare for release" commit.

Agree that this is not ideal.

> How about...

Seems to work on a local shallow clone, yes I can maybe try that. One drawback is we might fetch objects without the user knowing it if they run the script locally, but let's say it's acceptable if we print a message.

File: contrib/scripts/print-downgrade-version.sh
Comment (qmonnet) [reply contrib/scripts/print-downgrade-version.sh]: Thanks for the tip, I didn't know that! I kept the current form, as Paul says it's probably faster to understand for most people.

File: contrib/scripts/print-downgrade-version.sh
Comment (qmonnet) [reply contrib/scripts/print-downgrade-version.sh]: Done now

**Comment by qmonnet** [comment]
New approach: `git fetch` in the script to get the tag locally, as suggested by Maxim.

File: .github/workflows/tests-ipsec-upgrade.yaml
Comment (qmonnet) [inline .github/workflows/tests-ipsec-upgrade.yaml:162]: `github-sec` reviewers: Please double-check that this is OK!

File: .github/workflows/tests-ipsec-upgrade.yaml
Comment (gentoo-root) [inline .github/workflows/tests-ipsec-upgrade.yaml:202]: I'm not sure `persist-credentials: false` removes the previously stored ones. Did you check that?

Comment (qmonnet) [reply .github/workflows/tests-ipsec-upgrade.yaml:202]: My understanding is that it unsets the `http.https://github.com/.extraheader` Git config option used for passing the token with the requests (and the `core.sshCommand` as well, that we're not using as far as I understand), so I'd expect them to gone from the repo? I'm not sure how to check this for good :thinking: 

Screenshot_20240221_183127

Comment (gentoo-root) [reply .github/workflows/tests-ipsec-upgrade.yaml:202]: >it unsets

If so, that should do.

>how to check this for good

I would just try `git fetch` again after that step to see that now it fails. Not 100% robust, but still informative enough.

**Review by thorn3r — APPROVED** [review]

File: .github/workflows/tests-ipsec-upgrade.yaml
Comment (pchaigno) [reply .github/workflows/tests-ipsec-upgrade.yaml:202]: This change feels a bit risky, so I'd go as far as encoding the action suggested by Maxim in the workflow (fail if we can still fetch) so that we are notified of any change in behavior.

Comment (qmonnet) [reply .github/workflows/tests-ipsec-upgrade.yaml:202]: Good idea, I added it as a failsafe: https://github.com/cilium/cilium/pull/30742/files#diff-07b1303f71b74ecfe10ad34472da7c7e9b79ac9274fd93fe833ecc1551898473R204 and tested it successfully.

**Review by pchaigno — APPROVED** [review]

File: .github/workflows/tests-ipsec-upgrade.yaml
Comment (gentoo-root) [reply .github/workflows/tests-ipsec-upgrade.yaml:202]: Emphasizing that this check can return false positives in case of network errors (`git fetch` fails not because of missing creds, but because of something else, and we think the check passed), but eventually we should see it trigger if the credential removal behavior changes. It's not ideal, but definitely worth having.

We could try to analyze specific error codes returned by git, but it can return 128 for both kinds of errors.

**Review by gentoo-root — APPROVED** [review]

File: .github/workflows/tests-ipsec-upgrade.yaml
Comment (qmonnet) [reply .github/workflows/tests-ipsec-upgrade.yaml:202]: Except that, obviously, `git fetch` is _supposed_ to work on `cilium/cilium` where we don't need any auth to fetch the code :facepalm: thus triggering the failsafe on this PR. I need to implement this check differently.

Comment (qmonnet) [reply .github/workflows/tests-ipsec-upgrade.yaml:202]: Added a check as follows:

```yaml
      - name: Check we effectively removed Git credentials
        shell: bash
        run: |
          if ! curl -sL \
            -H "Accept: application/vnd.github+json" \
            -H "X-GitHub-Api-Version: 2022-11-28" \
            "https://api.github.com/repos/${{ github.repository }}" | \
            jq --exit-status '.private == false'; then
            echo 'Checking whether "git fetch" succeeds'
            if git fetch origin HEAD; then
              echo "::error::Git credentials not removed, aborting now."
              false
            fi
          fi
```

Comment (qmonnet) [reply .github/workflows/tests-ipsec-upgrade.yaml:202]: > Emphasizing that this check can return false positives in case of network errors

Sorry, I'm only seeing your comment now. Hmm yep, network errors would make the test “pass” as well. Not sure what we can do about it, maybe parsing the command output but that's a check on a check on a check and looks brittle...

My current implementation also has another limitation, which is that it doesn't run this check on a public repo (such as `cilium/cilium`), because I don't know how to test, without pushing stuff, that auth has been removed in that case :disappointed:

Comment (qmonnet) [reply .github/workflows/tests-ipsec-upgrade.yaml:202]: I discussed offline with Feroz. Not sure there's much to do for either case, the network errors or the workflow skipping the check on `cilium/cilium`. As far as I understand, the token we use only has limited permissions anyway (the only thing that regular public users couldn't do without it is setting the status for the workflow), so it seems sufficient to keep the check as is.

I'll go ahead and send to the merge queue so that patch releases for March don't get blocked on this.

Thanks all!

**Comment by pippolo84** [comment]
@qmonnet just a heads up that I was unable to backport this to v1.12 because of the conflicts, so I added the backport-author label.

**Pull request merged**
